### PR TITLE
Pet Daemon

### PR DESCRIPTION
**Pull request by kriskowal**
Individually reviewable commits.

This change will introduce pet-named workers, storage, and values to the Endo Daemon and expose them with its CLI. With this change, the following workflows become possible:

Ephemeral evaluation in an isolated worker:

```
$ endo spawn worker
$ endo eval worker 10
10
```

Persisted evaluation from a worker, using a pet name and a derivation path.

```
$ endo eval worker 10 --name ten
10
$ endo restart
$ end show ten
10
```

Evaluation with endowments:

```
$ endo eval worker 'ten + ten' ten --name twenty
20
$ endo restart
$ endo show twenty
20
```

Storage:

```
$ endo store <(echo hi) --name hi
$ endo cat hi
hi
```

Storage of archive:

```
$ endo archive foo.js --name foo
$ endo cat foo | tar t
... foo.js
```

These are a preamble to composite commands like `endo install`, which would store, spawn, and eval (or rather, importBundle), and subsequent work to reveal pet named capabilities to workers on-demand at the user’s discretion. `endo follow [name]` is also an obvious next step, for following a pet-named async iterator.

### CODE REVIEW

**Comment by dckc** [comment]
How do I run `endo spawn worker`? I don't see a `bin/endo`. I tried `yarn endo` without luck. halp?

**Comment by dckc** [comment]
more struggles:

```
~/projects/endo/packages/cli$ ./bin/endo store ./test/counter.js --name mkc
(TypeError#1)
TypeError#1: cmd.options is not a function

  at Command.<anonymous> (packages/cli/src/endo.js:217:28)
  at main (packages/cli/src/endo.js:314:19)
  at async packages/cli/bin/endo.cjs:4:3
```

fwiw, `./test/counter.js` contains:

```js
const makeCounter = () => {
  let value = 0;
  return {
    incr: () => (value += 1),
    decr: () => (value -= 1),
  };
};

makeCounter;
```

File: packages/daemon/src/daemon.js [hunk 1]
```diff
@@ -39,33 +44,149 @@ const sinkError = error => {
...
-const makeWorker = async locator => {
-  // @ts-ignore Node.js crypto does in fact have randomUUID.
-  const uuid = await crypto.randomUUID();
-  const workerCachePath = path.join(locator.cachePath, uuid);
-  await fs.promises.mkdir(workerCachePath, { recursive: true });
-  const logPath = path.join(workerCachePath, 'worker.log');
-  const output = fs.openSync(logPath, 'w');
-  const child = popen.fork(endoWorkerPath, [uuid, workerCachePath], {
-    stdio: ['ignore', output, output, 'pipe', 'ipc'],
+const provideReadableSha512 = (locator, sha512) => {
```
Comment (dckc) [inline packages/daemon/src/daemon.js:49]: as we discussed, seems to make a new one always. should be called `makeReadableSha512`

File: packages/cli/src/endo.js [hunk 2]
```diff
@@ -154,16 +172,139 @@ export const main = async rawArgs => {
...
+    });
+
+  program
+    .command('store <path>')
+    .option(
+      '-n,--name <name>',
+      'Assigns a pet name to the result for future reference',
+    )
+    .action(async (storablePath, cmd) => {
+      const { name } = cmd.options();
```
Comment (dckc) [inline packages/cli/src/endo.js:217]:
```suggestion
      const { name } = cmd.opts();
```

File: packages/daemon/src/daemon.js [hunk 3]
```diff
@@ -39,33 +44,149 @@ const sinkError = error => {
...
+  return Far(`Readable file with SHA-512 ${sha512.slice(0, 8)}...`, {
+    sha512() {
+      return sha512;
+    },
+    stream() {
```
Comment (dckc) [inline packages/daemon/src/daemon.js:54]: I added a `text()` method a la:

```diff
index c8b964637..42e6bbd71 100644
--- a/packages/daemon/src/daemon.js
+++ b/packages/daemon/src/daemon.js
@@ -47,13 +47,21 @@ const sinkError = error => {
  * @param {string} sha512
  */
 const provideReadableSha512 = (locator, sha512) => {
+  const getPath = () => {
+    const storageDirectoryPath = path.join(locator.statePath, 'store-sha512');
+    const storagePath = path.join(storageDirectoryPath, sha512);
+    return storagePath;
+  };
+
   return Far(`Readable file with SHA-512 ${sha512.slice(0, 8)}...`, {
     sha512() {
       return sha512;
     },
+    text() {
+      return fs.promises.readFile(getPath(), 'utf8');
+    },
     stream() {
-      const storageDirectoryPath = path.join(locator.statePath, 'store-sha512');
-      const storagePath = path.join(storageDirectoryPath, sha512);
+      const storagePath = getPath();
       const nodeReadStream = fs.createReadStream(storagePath);
       const reader = makeNodeReader(nodeReadStream);
       return makeReaderRef(reader);
~
```

**Comment by dckc** [comment]
I tried to `endo store` without first doing an `endo start`. The resulting diagnostic was not maximally informative:

`Error#1: connect ENOENT /run/user/1000/endo/captp0.sock`

**Comment by dckc** [comment]
Is there a missing `mkdir()` somewhere?

I kept getting `EACCES: permission denied /run/user/1000/endo/captp0.sock`. Even with `export XDG_RUNTIME_DIR=/tmp/r1`, I got `EACCES: permission denied /tmp/r1/endo/captp0.sock`.

`mkdir /tmp/r1/endo` seems to make it happy.

**Comment by dckc** [comment]
does this thing use sloppy eval? shouldn't this give `ReferenceError`?

```
$ ./bin/endo eval w1 'atob' mkc
undefined
```

**Review by dckc — COMMENTED** [review]
notes so far...

File: packages/daemon/src/daemon.js (hunk 3 above)
Comment (dckc) [reply packages/daemon/src/daemon.js:54]: that allowed me to do:

```
$ ./bin/endo eval w1 'E(mkc).text()' mkc
const makeCounter = () => {
  let value = 0;
  return {
    incr: () => (value += 1),
    decr: () => (value -= 1),
  };
};

makeCounter;
```

but this hangs:

```
./bin/endo eval w1 'E(mkc).text().then(eval)' mkc
```

I couldn't find a worker log. (I did find a worker .pid file.)

**Comment by dckc** [comment]
Is it as-designed that objects-as-closures don't keep their internal state across restarts?

This surprised me:

```
endo/packages/cli$ ./bin/endo eval w1 'E(c1).incr()' c1 --name five
5
endo/packages/cli$ ./bin/endo restart
endo/packages/cli$ ./bin/endo eval w1 'five' five
1
```

That was after...

```
endo/packages/cli$ ./bin/endo store ./test/counter.js --name csrc
endo/packages/cli$ ./bin/endo eval w1 'E(csrc).text().then(eval)' csrc --name mcc
Object [Alleged: makeCounter] {}
endo/packages/cli$ ./bin/endo eval w1 'E(mcc).make()' mcc
Object [Alleged: Counter] {}
endo/packages/cli$ ./bin/endo eval w1 'E(mcc).make()' mcc --name c1
Object [Alleged: Counter] {}
endo/packages/cli$ ./bin/endo eval w1 'E(c1).incr()' c1
1
endo/packages/cli$ ./bin/endo eval w1 'E(c1).incr()' c1
2
endo/packages/cli$ ./bin/endo eval w1 'E(c1).incr()' c1
3
endo/packages/cli$ ./bin/endo eval w1 'E(c1).incr()' c1
4
```

**Comment by kriskowal** [comment]
@Jack-Works May I solicit your assistance in debugging the Windows CI issue I’m encountering here? We are now running Endo tests on bare Windows in CI to make sure we continue to be sensitive to platform differences, using techniques you taught us (using `url.pathToFileURL` and `url.filePathToURL` for example) and now having learned the hard way that a Windows file handle can’t be safely used as a duplex connection. I’ve gotten the tests in `packages/daemon` (`yarn test`) to work locally on Windows, but this remaining CI failure is mysterious to me. Would you be willing to attempt to reproduce?

**Comment by Jack-Works** [comment]
```
~/../endo/../daemon master ≡  yarn test
yarn run v1.22.0
$ ava

  ✔ lifecycle (528ms)
  ─

  1 test passed
Done in 1.57s.
```

Is this correct? 👀

**Comment by kriskowal** [comment]
> ```
> ~/../endo/../daemon master ≡  yarn test
> yarn run v1.22.0
> $ ava
> 
>   ✔ lifecycle (528ms)
>   ─
> 
>   1 test passed
> Done in 1.57s.
> ```
> 
> Is this correct? 👀

That’s correct on `master`, can you check the branch `kris-pet-daemon`? There are some new tests that are failing in CI, around getting pet names to work in an `endo` environment executor.

**Comment by kriskowal** [comment]
> more struggles:
> 
> ```
> ~/projects/endo/packages/cli$ ./bin/endo store ./test/counter.js --name mkc
> (TypeError#1)
> TypeError#1: cmd.options is not a function
> 
>   at Command.<anonymous> (packages/cli/src/endo.js:217:28)
>   at main (packages/cli/src/endo.js:314:19)
>   at async packages/cli/bin/endo.cjs:4:3
> ```

Addressing in https://github.com/endojs/endo/pull/1471

**Comment by kriskowal** [comment]
> > ```
> > ~/../endo/../daemon master ≡  yarn test
> > yarn run v1.22.0
> > $ ava
> > 
> >   ✔ lifecycle (528ms)
> >   ─
> > 
> >   1 test passed
> > Done in 1.57s.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Is this correct? 👀
> 
> That’s correct on `master`, can you check the branch `kris-pet-daemon`? There are some new tests that are failing in CI, around getting pet names to work in an `endo` environment executor.

Thank you for looking into this @Jack-Works. I discovered that the failures on Windows impacted all tests on this branch, not just the new ones. By bisecting and running these tests in a CI environment (the only environment sufficiently resource constrained that the child process fails to completely flush CapTP after its own termination), I was able to address the underlying problem and get all Windows tests passing by ignoring the exception that propagates through the `closed` promise from CapTP. There is probably a better solution for graceful shutdown, but this one’s mine.

**Review by dckc — APPROVED** [review]
it's pretty useful

I'm kinda used up as a reviewer. Another set of eyes would be good, if we can afford it.

**Comment by kriskowal** [comment]
All feedback captured in the `endo` branch. Thank you!

**Pull request closed without merging**